### ISSUE TITLE
[DumpUnitary] Reimplement the tool using DumpOperation

### ISSUE THREAD

**Issue by tcNickolas** [microsoft/QuantumKatas#499]
[DumpUnitary](https://github.com/microsoft/QuantumKatas/tree/main/utilities/DumpUnitary) is a tool that accompanies the [UnitaryPatterns kata](https://github.com/microsoft/QuantumKatas/tree/master/UnitaryPatterns) - it allows the learner to dump the unitary implemented by an operation and extracts the "pattern" of zeros and non-zeros according to the rules defined in the kata. It predates the recently added [DumpOperation](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.diagnostics.dumpoperation) tool, so it uses a different, less elegant and less accurate approach. In particular, the columns of the matrix are obtained independently, and the relative phase between the columns is not preserved. Thus the matrix dumped might differ from the actual unitary implemented in phases of the entries.

It makes sense to preserve the tool (it's not made completely obsolete by DumpOperation, since it allows to convert the matrix to a pattern), but to replace the parts that use DumpMachine to dump column elements to a series of files, read the files and assemble the matrices from them with one call to DumpOperation and reading the resulting file.

**Comment by cgranade** [comment]
> It makes sense to preserve the tool (it's not made completely obsolete by DumpOperation, since it allows to convert the matrix to a pattern), but to replace the parts that use DumpMachine to dump column elements to a series of files, read the files and assemble the matrices from them with one call to DumpOperation and reading the resulting file.

You should be able to use the  `OnDisplayableDiagnostic` event of the `QuantumSimulator` class to directly get the `DisplayableUnitary` object emitted by `DumpOperation` without needing to write to a file and read it back. Happy to point to that if it would be helpful. Thanks!

**Comment by tcNickolas** [comment]
Yes, please do - it can be tricky to locate which repo the specific functionality resides at.

**Comment by cgranade** [comment]
This is the event where displayable diagnostics are emitted: https://github.com/microsoft/qsharp-runtime/blob/main/src/Simulation/Common/SimulatorBase.cs#L60
This is the type of the diagnostic object corresponding to the output emitted by `DumpOperation`: https://github.com/microsoft/QuantumLibraries/blob/main/Standard/src/Diagnostics/Emulation/DataStructures.cs#L19

You can use the two together to catch `DisplayableUnitaryOperator` objects emitted to the diagnostic pipeline:

```C#
var qsim = new QuantumSimulator();
qsim.OnDisplayableDiagnostic += (diagnostic) =>
{
    if (diagnostic is DisplayableUnitaryOperator op)
    {
         // do something with it here...
    }
};
```

**Comment by jainvasu631** [comment]
@tcNickolas 
I remember using this tool while writing and debugging the Unitary Patterns Kata. I can work on this, unless you or someone else is already working on it.  I would love to check out the DumpOperation utility which was so needed, requested and awaited before it came out.

**Comment by tcNickolas** [comment]
@jainvasu631 Apologies for missing your comment - sure, I don't think anybody is working on this issue, so you're welcome to it :-)
